### PR TITLE
Fix HTTP responses in 4xx - 5xx range in call mediator blocking mode

### PR DESCRIPTION
**Pull request by Vathsan**
## Purpose
Fixes [product-ei/issues/951](https://github.com/wso2/product-ei/issues/951#issue-253922454)

Behaviour before PR: 
When Call Mediator is used in mediation flow with blocking enabled if the endpoint responds non-success response (ex: 401, 403) it goes to error sequence. This issue occurs because in axis2. transport (handleResponse method of HTTPSender class), the default behaviour of handling HTTP status code which is not in SUCCESSFUL HTTPStatusCodeFamily is to throw new AxisFault (If HTTP status code is 400, 500 0r 409 this won’t affect).

Behaviour after fix/PR: 
Both success and non-success response codes will be sent back to the client instead of throwing AxisFault as of non-blocking mode.

### CODE REVIEW

File: modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java [hunk 1]
```diff
@@ -323,45 +322,12 @@ private void handleResponse(MessageContext msgContext,
                 }
             }
         }
-        if (HTTPStatusCodeFamily.SUCCESSFUL.equals(family)) {
+        if (!errorCodes.contains(statusCode)) {
             // Save the HttpMethod so that we can release the connection when cleaning up
             // HTTP 202 Accepted should support body based on ESBJAVA-4370 as spec does not strictly enforce
             // no-entity-body with 202 Accepted response.
```
Comment (a5anka) [inline modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java]: Is the above comment still valid?

**Comment by a5anka** [comment]
@Vathsan Can you add information on the before and after behaviour of the fix to the PR?

**Comment by a5anka** [comment]
@Vathsan Shall we check the behaviour in message processor with the fix?

File: modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java (hunk 1 above)
Comment (Vathsan) [reply modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java]: Removed the comment in [8146236](https://github.com/wso2/wso2-axis2/pull/166/commits/814623632e32e24d87c1195971fd45302eaeaace) as it becomes invalid after the fix.

File: modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java [hunk 2]
```diff
@@ -323,45 +322,9 @@ private void handleResponse(MessageContext msgContext,
...
-            String value = null;
-            if (contenttypeHeader != null) {
-                value = contenttypeHeader.getValue();
-            }
-             OperationContext opContext = msgContext.getOperationContext();
-            if(opContext!=null){
-                MessageContext inMessageContext =
-                        opContext.getMessageContext(WSDLConstants.MESSAGE_LABEL_IN_VALUE);
-                if(inMessageContext!=null){
-                    inMessageContext.setProcessingFault(true);
```
Comment (Nirothipan) [inline modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java]: Did we check the effect of removing this line ?

**Comment by Nirothipan** [comment]
@Vathsan since we are removing nonhttp error codes with it , create a doc issue also for this

File: modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java [hunk 3]
```diff
@@ -310,7 +310,6 @@ private void handleResponse(MessageContext msgContext,
         int statusCode = method.getStatusCode();
         HTTPStatusCodeFamily family = getHTTPStatusCodeFamily(statusCode);
```
Comment (Nirothipan) [inline modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java]: do we need this anymore ? and the method getHTTPStatusCodeFamily ?

**Comment by Vathsan** [comment]
> @Vathsan Can you add information on the before and after behaviour of the fix to the PR?

Done

File: modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java (hunk 3 above)
Comment (Vathsan) [reply modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java]: We do not need this line and the method after the fix. Removed them both in commit cdb42256f6841648687e25969304e19eb24e7591

File: modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java (hunk 2 above)
Comment (Vathsan) [reply modules/transport/http/src/org/apache/axis2/transport/http/HTTPSender.java]: I ran mediation profile test cases with the fix and made sure there were no test failures.

**Comment by Vathsan** [comment]
> @Vathsan Shall we check the behaviour in message processor with the fix?

After the fix message processor will not attempt to retry if it gets a non-success response from the backend. But we can make it retry if it receives a non-success response from backend by setting the particular status code in `error.http.status.codes` property. 
As we have also removed `non.error.http.status.codes` property, we do not need to set Non-retry HTTP status codes in message processor.

**Comment by Vathsan** [comment]
> @Vathsan since we are removing nonhttp error codes with it , please create a doc issue also for this

Related doc issue can be found in [product-ei/issues/3723](https://github.com/wso2/product-ei/issues/3723)

**Comment by isudana** [comment]
We need to hold this change as this will affect a wider audience than just Call mediator and Message processor. 
Also, we need to invest some cycles to see whether we can use a single transport implementation for both blocking and non-blocking behaviours.
That will be the most reliable solution to address the inconsistencies in blocking vs non-blocking behaviours.

**Comment by sreenath073** [comment]
Hi @Vathsan 
I am using wso2 ESB for implementing my work flows. I am developing using wso2 Integration Studio 7.0 version, I see this inconsistency while using call mediator in blocking mode and non-blocking mode.
In blocking mode, I am not able to catch the individual http codes like 401,403,501,503 etc and when the call fails, the control directly goes to the fault sequence defined.
To overcome this problem, I switched to non-blocking mode, but my mediation flow completes and I am able to catch the exception, but surprisingly, I see one particular error in the logs,
[2020-06-30 01:22:41,966] ERROR {org.apache.synapse.mediators.builtin.PropertyMediator} - Unknown type : INTEGER for the property mediator or the property value cannot be converted into the specified type. java.lang.NumberFormatException: For input string: ""

Can you please help me resolve this issue, either in blocking mode or non-blocking mode. 
Basically I am trying to catch individual http error codes for better error handling in my flows.
Thanks
Sreenath
